### PR TITLE
raft_client: fix imbalance load of raft msg

### PR DESCRIPTION
**Pull request by gengliqi**
Signed-off-by: gengliqi <gengliqiii@gmail.com>

### What problem does this PR solve?

Problem Summary:

1. imbalance load of raft msg: raft client uses `region_id % grpc-raft-conn-num` to choose the specific connection. It's no problem when `grpc-raft-conn-num` is 1(default value). However, when `grpc-raft-conn-num` is larger, for example, 2, the distribution maybe not uniform because of the region id. If the num of TiKV is 3, the region id maybe always even because PD allocates region id with 3 peer id. To fix this problem, @5kbpers suggests using seahash. I test it with a parity check and find the distribution is very uniform. Even if the value is even, its hash value is odd or even. Also, the speed of seahash is good enough(benchmark result shows 11ns/iter on my laptop).
~~2. raft msg flush strategy: The raft msg is flushed only after handling all regions msg of this round. It may introduce some unnecessary latency. However, if the raft msg is flushed every time, the grpc CPU will be very high because of a large number of notifying. To alleviate this problem, the delay time is introduced. But I think this method is not very elegant. Maybe we should refactor this logic later.~~

### What is changed and how it works?

What's Changed:
explained above.

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- No code

Side effects

- No

### Release note 
* No Release Note

### CODE REVIEW

**Comment by BusyJay** [comment]
>  It may introduce some unnecessary latency.

Do you have any stats to show how large is the latency?

File: components/raftstore/src/store/fsm/store.rs [hunk 1]
```diff
@@ -681,6 +681,7 @@ impl<EK: KvEngine, ER: RaftEngine, T: Transport> RaftPoller<EK, ER, T> {
                 PeerFsmDelegate::new(&mut peers[ready.batch_offset], &mut self.poll_ctx)
                     .post_raft_ready_append(ready);
             }
+            self.poll_ctx.trans.try_flush();
```
Comment (hicqu) [inline components/raftstore/src/store/fsm/store.rs:684]: Is it compatible with your origin design? @BusyJay

**Comment by hicqu** [comment]
Maybe it's necessary to do a benchmark for the new flush mechanism. @gengliqi maybe your new mechanism can be used with https://github.com/tikv/grpc-rs/pull/469?

**Comment by zhangjinpeng1987** [comment]
Better to split this PR into two, each one solves an individual issue.

File: src/server/raft_client.rs [hunk 2]
```diff
@@ -844,7 +845,20 @@ where
...
+            if self.last_hash.0 == 0 || msg.region_id != self.last_hash.0 {
+                let mut hasher = SeaHasher::new();
+                hasher.write_u64(msg.region_id);
+                let h = hasher.finish();
+                self.last_hash = (
+                    msg.region_id,
+                    h % self.builder.cfg.grpc_raft_conn_num as u64,
+                );
+            };
+            self.last_hash.1 as usize
```
Comment (BusyJay) [inline src/server/raft_client.rs:860]: I'm afraid this will introduce out of order messages.

Comment (gengliqi) [reply src/server/raft_client.rs:860]: Why? I think it is the same as the former logic that the same `region_id` corresponds to the same `conn_id`.

File: tests/integrations/config/test-custom.toml [hunk 3]
```diff
@@ -43,6 +43,10 @@ status-addr = "example.com:443"
 advertise-status-addr = "example.com:443"
 status-thread-pool-size = 1
 max-grpc-send-msg-len = 6291456
+raft-client-grpc-send-msg-buffer = 1263616
+raft-client-queue-size = 1234
+raft-msg-max-batch-size = 123
+raft-msg-flush-delay-us = 1234
```
Comment (BusyJay) [inline tests/integrations/config/test-custom.toml:49]: Stale change.

File: src/server/raft_client.rs [hunk 4]
```diff
@@ -844,7 +845,20 @@ where
     /// are sent out.
     pub fn send(&mut self, msg: RaftMessage) -> result::Result<(), DiscardReason> {
         let store_id = msg.get_to_peer().store_id;
-        let conn_id = (msg.region_id % self.builder.cfg.grpc_raft_conn_num as u64) as usize;
+        let conn_id = if self.builder.cfg.grpc_raft_conn_num == 1 {
+            0
+        } else {
+            if self.last_hash.0 == 0 || msg.region_id != self.last_hash.0 {
+                let mut hasher = SeaHasher::new();
```
Comment (BusyJay) [inline src/server/raft_client.rs:852]: You can use `seahash::hash(msg.region_id.as_ne_bytes())` to get better performance.

Comment (gengliqi) [reply src/server/raft_client.rs]: Done.

**Review by BusyJay — APPROVED** [review]

**Review by zhangjinpeng1987 — APPROVED** [review]
LGTM

**Pull request merged**